Query for the relative positions of multiple Hidden Markov model hits across multiple genomes in HMMER-DB.
==========================================================================================================
```{r message = FALSE}
# Imports:
library(ggplot2)
library(RSQLite)
```

#### We can query the HMMER-DB database using a SQL query within R. Our first step is to setup a database connection:
```{r}
# Setting up database connection:
sqlite = dbDriver("SQLite")
HMMDB  = dbConnect(sqlite, "/Users/lee/Data/SteriodHMMs/OldDBs/HMMDBV4.sqlite") # Location of HMMER-DB Sqlite database.
```
#### The code below queries the database and loads the results of the query into a R dataframe (Note: This is a rather complex query with multiple filters). 
```{r}
# Executes SQL query and loads results directly into a dataframe. 
data = dbGetQuery(HMMDB, "/* SQL Outer Query: Wrapper for subquery one that calculates relative protein center from subquery one's relative protein start and stop. */
SELECT DISTINCT
  HMM_Data.HMM_Family,
	Organisms.Organism_Description,
	(( CAST (Proteins.'End' AS REAL) + CAST (Proteins.Start AS REAL)) / 2 ) AS Position
FROM
	HMM_Data,
	HMM_Hits,
	Organisms,
	Proteins
WHERE
	HMM_Data.HMM_Model = HMM_Hits.HMM_Model
AND HMM_Hits.Protein_Accession = Proteins.Protein_Accession
AND Proteins.Organism_Accession = Organisms.Organism_Accession
AND Organisms.Organism_Accession IN (
	/* SQL Inner Query 1: Selects organisms from the subquery with a HMM_Coverage of at least 60% for key steriod degrading enzymes.*/
	SELECT
		Organisms.Organism_Accession
	FROM
		HMM_Data,
		HMM_Hits,
		Organisms,
		Proteins
	WHERE
		HMM_Data.HMM_Model = HMM_Hits.HMM_Model
	AND HMM_Hits.Protein_Accession = Proteins.Protein_Accession
	AND Proteins.Organism_Accession = Organisms.Organism_Accession
	AND HMM_Data.HMM_Family IN (
		'hsaA',
		'hsaC',
		'hsaD',
		'hsaE',
		'hsaF',
		'hsaG',
		'hsd4B',
		'KshA',
		'kstD'
	)
	AND HMM_Hits.HMM_Coverage >= 0.8
	AND Organisms.Organism_Accession IN (
		/* SQL Inner Query 2: Selects only organisms that hsaC hits with over 80% HMM coverage from organism with KshA*/
		SELECT DISTINCT
			Organisms.Organism_Accession
		FROM
			HMM_Hits,
			Organisms,
			Proteins
		WHERE
			HMM_Hits.Protein_Accession = Proteins.Protein_Accession
		AND Organisms.Organism_Accession = Proteins.Organism_Accession
		AND HMM_Hits.HMM_Model LIKE 'hsaC%' /* Selects organisms with a hit for HsaC*/
		AND HMM_Hits.HMM_Coverage >= 0.8 /* Select only hsaC hits with over 80% HMM coverage */
		AND Organisms.Organism_Accession IN (
			/* SQL Inner Query 3: Selects only organisms that KshA hits with over 80% HMM coverage.*/
			SELECT DISTINCT
				Organisms.Organism_Accession
			FROM
				HMM_Hits,
				Organisms,
				Proteins
			WHERE
				HMM_Hits.Protein_Accession = Proteins.Protein_Accession
			AND Organisms.Organism_Accession = Proteins.Organism_Accession
			AND HMM_Hits.HMM_Model LIKE 'KshA%' /* Selects organisms with a hit for KshA*/
			AND HMM_Hits.HMM_Coverage >= 0.8
			AND Organisms.Organism_Description LIKE 'Myco%' /* Selects non-Mycbacterium */
		)
	)
	GROUP BY
		Organisms.Organism_Accession
	HAVING
		count(DISTINCT HMM_Data.HMM_Family) >= 5.4
	ORDER BY
		Organisms.Organism_Phylogeny ASC
)
AND HMM_Data.HMM_Family IN (
	'hsaA',
	'hsaC',
	'hsaD',
	'hsaE',
	'hsaF',
	'hsaG',
	'hsd4B',
	'KshA',
	'kstD'
)
ORDER BY
	Organisms.Organism_Phylogeny ASC,
	Organisms.Organism_Description ASC,
	Position ASC")
```

#### We now check if the SQL query is generating the dataframe we want:
```{r}
head(data) # Note: The final column may wrap to the next line.
```

#### We can then use ggplot to plot the relative positions of HMM hits in the genomes of the organisms in the database:
```{r fig.width = 15, fig.height = 10}
plotObj = ggplot(data, aes(x = Position, y = Organism_Description, color = factor(HMM_Family)))
plotObj + geom_point(alpha = 3/4) + # Slight alpha so one can visualize overlaping points better.
          ggtitle("Positions of proteins with HMM hits for Mycobacterium in the database.") + 
          xlim(0, 10000000) +
          xlab("Protein's postion on chromosome/plasmid  (bp).") + ylab("Organism") + labs(colour = "Type of HMM hit:") 
```

*Please note that most bacterial genomes are circular and the genome may start at an arbitrary point.*

#### We can also use ggplot to plot the relative density of HMM hits in the genomes of the organisms in the database:
```{r fig.width = 15, fig.height = 10}
plotObj = ggplot(data, aes(x = Position, y = Organism_Description, fill = 'black'))
plotObj + geom_point(alpha = 1/5) + # Proteins are all the same color and we use alpha to highlight areas with a greater number of hit proteins.
          ggtitle("Density of proteins with HMM hits for Mycobacterium in the database.") + 
          xlim(0, 10000000) +
          xlab("Protein's postion on chromosome/plasmid  (bp).") + ylab("Organism") + theme(legend.position = "none")
```

*Please note that most bacterial genomes are circular and the genome may start at an arbitrary point.*

#### Lets say we want to filter the hits down further to just the key enzymes such as HsaC and KshA. We can use R to make a new filtred dataframe.
```{r}
dataTwo = data[data$HMM_Family %in% c("KshA", "hsaC"), ] # Select rows where the column HMM_Family is equal to KshA or hsaC.
```
#### We now check if the new dataframe is what we want:
```{r}
head(dataTwo) # Note: The final column may wrap to the next line.
```

#### We again use ggplot to plot the relative positions of HMM hits in the genomes of the organisms in the database but now only for HsaC and KshA:
```{r fig.width = 15, fig.height = 10}
plotObj = ggplot(dataTwo, aes(x = Position, y = Organism_Description, color = factor(HMM_Family)))
plotObj + geom_point(alpha = 3/4) + # Slight alpha so one can visualize overlaping points better.
          ggtitle("Positions of HsaC and KshA HMM hits for Mycobacterium in the database.") + 
          xlim(0, 10000000) +
          xlab("Protein's postion on chromosome/plasmid (bp).") + ylab("Organism") + labs(colour = "Type of HMM hit:") 
```
